docs/Config.md
# Config

This document describes configuration of Open-VE.

You can configure Open-VE by CLI flags, environment variables, or config file.

## CLI Flags or Environment Variables

| CLI Args                             | Env                                        | Default      | Desc                                                                                   |
| ------------------------------------ | ------------------------------------------ | ------------ | -------------------------------------------------------------------------------------- |
| `--mode`                             | `OPEN-VE_MODE`                             | `master`     | master or slave                                                                        |
| `--slave-id`                         | `OPEN-VE_SLAVE_ID`                         |              | Unique slave ID (if mode is slave, this is required)                                   |
| `--slave-slave-http-addr`            | `OPEN-VE_SLAVE_SLAVE_HTTP_ADDR`            |              | HTTP server address (if mode is slave, this is required)                               |
| `--slave-master-http-addr`           | `OPEN-VE_SLAVE_MASTER_HTTP_ADDR`           |              | Master HTTP server address (if mode is slave, this is required)                        |
| `--slave-master-authn-method`        | `OPEN-VE_SLAVE_MASTER_AUTHN_METHOD`        | `none`       | Authentication method of the master server (preshared)                                 |
| `--slave-master-authn-preshared-key` | `OPEN-VE_SLAVE_MASTER_AUTHN_PRESHARED_KEY` |              | Preshared key of the master server (if authn method of the master server is preshared) |
| `--http-port`                        | `OPEN-VE_HTTP_PORT`                        | `8080`       | HTTP server port number                                                                |
| `--http-cors-allowed-origins`        | `OPEN-VE_HTTP_CORS_ALLOWED_ORIGINS`        | `["*"]`      | CORS allowed origins                                                                   |
| `--http-cors-allowed-headers`        | `OPEN-VE_HTTP_CORS_ALLOWED_HEADERS`        | `["*"]`      | CORS allowed headers                                                                   |
| `--http-tls-enabled`                 | `OPEN-VE_HTTP_TLS_ENABLED`                 | `false`      | HTTP server TLS enabled                                                                |
| `--http-tls-cert-path`               | `OPEN-VE_HTTP_TLS_CERT_PATH`               |              | HTTP server TLS cert path                                                              |
| `--http-tls-key-path`                | `OPEN-VE_HTTP_TLS_KEY_PATH`                |              | HTTP server TLS key path                                                               |
| `--grpc-port`                        | `OPEN-VE_GRPC_ADDR`                        | `9000`       | gRPC server port number                                                                |
| `--grpc-tls-enabled`                 | `OPEN-VE_GRPC_TLS_ENABLED`                 | `false`      | gRPC server TLS enabled                                                                |
| `--grpc-tls-cert-path`               | `OPEN-VE_GRPC_TLS_CERT_PATH`               |              | gRPC server TLS cert path                                                              |
| `--grpc-tls-key-path`                | `OPEN-VE_GRPC_TLS_KEY_PATH`                |              | gRPC server TLS key path                                                               |
| `--store-engine`                     | `OPEN-VE_STORE_ENGINE`                     | `memory`     | store engine (redis/memory)                                                            |
| `--store-redis-addr`                 | `OPEN-VE_STORE_REDIS_ADDR`                 | `redis:6379` | Redis address                                                                          |
| `--store-redis-password`             | `OPEN-VE_STORE_REDIS_PASSWORD`             |              | Redis password                                                                         |
| `--store-redis-db`                   | `OPEN-VE_STORE_REDIS_DB`                   | `0`          | Redis DB                                                                               |
| `--store-redis-pool-size`            | `OPEN-VE_STORE_REDIS_POOL_SIZE`            | `1000`       | Redis pool size                                                                        |
| `--log-level`                        | `OPEN-VE_LOG_LEVEL`                        | `info`       | Log level                                                                              |
| `--authn-method`                     | `OPEN-VE_AUTHN_METHOD`                     | `none`       | Authentication method of the server (preshared)                                        |
| `--authn-preshared-key`              | `OPEN-VE_AUTHN_PRESHARED_KEY`              |              | Preshared key of the server (if authn method is preshared)                             |

## Config File

Place the `config.yaml` in the same directory or `$HOME/.open-ve/config.yaml`.

```yaml
mode: "master"
http:
  port: "8080"
  corsAllowedOrigins: ["*"]
  corsAllowedHeaders: ["*"]
  tls:
    enabled: false
    certPath: ""
    keyPath: ""
grpc:
  poer: "9000"
  tls:
    enabled: false
    certPath: ""
    keyPath: ""
store:
  engine: "redis" # redis or memory
  redis:
    addr: "redis:6379"
    password: ""
    db: 0
    poolSize: 1000
log:
  level: "info"
```


go/pkg/config/config.go
package config

type Config struct {
	Mode  string      `yaml:"mode"`
	Slave SlaveConfig `yaml:"slave"`
	Http  HttpConfig  `yaml:"http"`
	GRPC  GRPCConfig  `yaml:"grpc"`
	Store StoreConfig `yaml:"store"`
	Log   LogConfig   `yaml:"log"`
	Authn AuthnConfig `yaml:"authn"`
}

type SlaveConfig struct {
	Id             string      `yaml:"id"`
	SlaveHTTPAddr  string      `yaml:"slaveHTTPAddr"`
	MasterHTTPAddr string      `yaml:"masterHTTPAddr"`
	MasterAuthn    AuthnConfig `yaml:"masterAuthn"`
}

type HttpConfig struct {
	Port               string    `yaml:"port"`
	CORSAllowedOrigins []string  `yaml:"corsAllowedOrigins"`
	CORSAllowedHeaders []string  `yaml:"corsAllowedHeaders"`
	TLS                TLSConfig `yaml:"tls"`
}

type GRPCConfig struct {
	Port string    `yaml:"port"`
	TLS  TLSConfig `yaml:"tls"`
}

type RedisConfig struct {
	Addr     string `yaml:"addr"`
	Password string `yaml:"password" json:"-"`
	DB       int    `yaml:"db"`
	PoolSize int    `yaml:"poolSize"`
}

type StoreConfig struct {
	Engine string      `yaml:"engine"`
	Redis  RedisConfig `yaml:"redis"`
}

type LogConfig struct {
	Level string `yaml:"level"`
}

type TLSConfig struct {
	Enabled  bool   `yaml:"enabled"`
	CertPath string `yaml:"certPath"`
	KeyPath  string `yaml:"keyPath"`
}

type AuthnConfig struct {
	Method    string          `yaml:"method"`
	Preshared PresharedConfig `yaml:"preshared"`
}

type PresharedConfig struct {
	Key string `yaml:"key"`
}

func DefaultConfig() *Config {
	return &Config{
		Mode: "master",
		Slave: SlaveConfig{
			Id:             "",
			SlaveHTTPAddr:  "",
			MasterHTTPAddr: "",
			MasterAuthn: AuthnConfig{
				Method: "none",
				Preshared: PresharedConfig{
					Key: "",
				},
			},
		},
		Http: HttpConfig{
			Port:               "8080",
			CORSAllowedOrigins: []string{"*"},
			CORSAllowedHeaders: []string{"*"},
			TLS: TLSConfig{
				Enabled: false,
			},
		},
		GRPC: GRPCConfig{
			Port: "9000",
			TLS: TLSConfig{
				Enabled: false,
			},
		},
		Store: StoreConfig{
			Engine: "memory",
			Redis: RedisConfig{
				Addr:     "redis:6379",
				Password: "",
				DB:       0,
				PoolSize: 1000,
			},
		},
		Log: LogConfig{
			Level: "info",
		},
		Authn: AuthnConfig{
			Method: "none",
			Preshared: PresharedConfig{
				Key: "",
			},
		},
	}
}
